File: Battleship/src/main/java/de/htwg/battleship/actor/childactor/grandchild/BorderActor.java
// BorderController.java

package de.htwg.battleship.actor.childactor.grandchild;

import de.htwg.battleship.actor.messages.BorderMessage;
import de.htwg.battleship.model.IShip;

import static de.htwg.battleship.util.StatCollection.isBetween;

/**
 * BorderController which checks of the ship is in the field. use the
 * Chain-of-Responsibility-Pattern
 *
 * @author Moritz Sauter (SauterMoritz@gmx.de)
 * @version 1.00
 * @since 2014-12-15
 */

abstract class BorderActor extends AbstractGrandchildActor {
    static final String ACTOR_NAME = "borderer";

    @Override
    public void onReceive(final Object message) throws Exception {
        if (!(message instanceof BorderMessage)) {
            logUnhandled(ACTOR_NAME, message);
            return;
        }
        handleBorderMessage((BorderMessage) message);
    }

    abstract void handleBorderMessage(BorderMessage message);
}

/**
 * BorderController implementation for the false ship-orientation.
 *
 * @author Moritz Sauter (SauterMoritz@gmx.de)
 */
class BorderFalseActor extends BorderActor {
    private boolean isIn(final IShip ship, int boardSize) {
        int ylow = ship.getY();
        int yupp = ylow + ship.getSize() - 1;
        int x = ship.getX();
        return isBetween(boardSize - 1, 0, ylow) &&
               isBetween(boardSize - 1, 0, yupp) &&
               isBetween(boardSize - 1, 0, x);
    }

    @Override
    void handleBorderMessage(BorderMessage borderMessage) {
        getSender()
            .tell(isIn(borderMessage.getShip(), borderMessage.getBoardSize()),
                  getSelf());
    }
}

File: Battleship/src/main/scala/de/htwg/battleship/actor/BorderActor.scala
package de.htwg.battleship.actor

import akka.actor.Actor
import de.htwg.battleship.actor.messages.BorderMessage
import de.htwg.battleship.model.IShip
import de.htwg.battleship.util.StatCollection.isBetween

/**
  * BorderActor
  *
  * @author ms
  * @since 2016-06-16
  */

object BorderActor {
    val ACTOR_NAME = "borderer"
}

class BorderActor extends Actor {

    def isOnField(ship: IShip, boardSize: Int): Boolean = {
        val realBoardSize = boardSize - 1
        val low = if (ship.isOrientation) ship.getX else ship.getY
        val fix = if (ship.isOrientation) ship.getY else ship.getX
        val up = low + ship.getSize
        isBetween(realBoardSize, 0, low) &&
            isBetween(realBoardSize, 0, up) &&
            isBetween(realBoardSize, 0, fix)
    }

    /**
      * Method to receive messages.
      *
      * @return sends true to the asker if the ship is NOT on the field, sends false if the ship is placed on the board in a valid way
      */
    override def receive: Receive = {
        case msg: BorderMessage => sender() ! !isOnField(msg.ship, msg.boardSize)
        case msg => unhandled(msg)
    }
}
